src/providers/ogr/qgsogrprovider.cpp

```diff
@@ -634,130 +620,24 @@ QString QgsOgrProvider::ogrWkbGeometryTypeName( OGRwkbGeometryType type ) const
 {
-  QgsDebugMsg( "Entered." );
   if ( !mValid )
   {
     return QStringList();
   }
-
   if ( !mSubLayerList.isEmpty() )
-    return mSubLayerList;
-
-  for ( unsigned int i = 0; i < layerCount() ; i++ )
   {
-    OGRLayerH layer = OGR_DS_GetLayer( ogrDataSource, i );
-    OGRFeatureDefnH fdef = OGR_L_GetLayerDefn( layer );
-    QString layerName = QString::fromUtf8( OGR_FD_GetName( fdef ) );
-    OGRwkbGeometryType layerGeomType = OGR_FD_GetGeomType( fdef );
-
-    // ignore this layer if a sublayer was requested and it is not this one
-    if ( mIsSubLayer &&
-         ( ( !mLayerName.isNull() && layerName != mLayerName ) ||
-           ( mLayerName.isNull() && mLayerIndex >= 0 && i != ( unsigned int )mLayerIndex ) ) )
-    {
-      QgsDebugMsg( QString( "subLayers() ignoring layer #%1 (%2)" ).arg( i ).arg( layerName ) );
-      continue;
-    }
-
-    if ( !mIsSubLayer && ( layerName == QLatin1String( "layer_styles" ) ||
-                           layerName == QLatin1String( "qgis_projects" ) ) )
-    {
-      // Ignore layer_styles (coming from QGIS styling support) and
-      // qgis_projects (coming from http://plugins.qgis.org/plugins/QgisGeopackage/)
-      continue;
-    }
-
-    QgsDebugMsg( QString( "id = %1 name = %2 layerGeomType = %3" ).arg( i ).arg( layerName ).arg( layerGeomType ) );
-
-    if ( wkbFlatten( layerGeomType ) != wkbUnknown )
-    {
-      int layerFeatureCount = OGR_L_GetFeatureCount( layer, 0 );
-
-      QString geom = ogrWkbGeometryTypeName( layerGeomType );
-
-      mSubLayerList << QStringLiteral( "%1:%2:%3:%4" ).arg( i ).arg( layerName, layerFeatureCount == -1 ? tr( "Unknown" ) : QString::number( layerFeatureCount ), geom );
-    }
-    else
-    {
-      QgsDebugMsg( "Unknown geometry type, count features for each geometry type" );
-      // Add virtual sublayers for supported geometry types if layer type is unknown
-      // Count features for geometry types
-      QMap<OGRwkbGeometryType, int> fCount;
-      // TODO: avoid reading attributes, setRelevantFields cannot be called here because it is not constant
-      //setRelevantFields( ogrLayer, true, QgsAttributeList() );
-      OGR_L_ResetReading( layer );
-      OGRFeatureH fet;
-      while ( ( fet = OGR_L_GetNextFeature( layer ) ) )
-      {
-        OGRGeometryH geom = OGR_F_GetGeometryRef( fet );
-        if ( geom )
-        {
-          OGRwkbGeometryType gType = ogrWkbSingleFlatten( OGR_G_GetGeometryType( geom ) );
-          fCount[gType] = fCount.value( gType ) + 1;
-        }
-        OGR_F_Destroy( fet );
-      }
-      OGR_L_ResetReading( layer );
-      // it may happen that there are no features in the layer, in that case add unknown type
-      // to show to user that the layer exists but it is empty
-      if ( fCount.isEmpty() )
-      {
-        fCount[wkbUnknown] = 0;
-      }
-
-      // When there are CurvePolygons, promote Polygons
-      if ( fCount.contains( wkbPolygon ) && fCount.contains( wkbCurvePolygon ) )
```

Comment:
I never found a condition where this was called. One of the tests (test_02_OgrSpatialTableXYZM) to try to bring about this situation.
Since those geometries are being rendered properly, this may not be needed.

If any further "promotion" logic is needed, it should be done in:
* QgsOgrProviderUtils::getOgrGeomType

which is intended the central place (used by other Ogr classes) to determin the Geometry type.